Mathlib/Analysis/Normed/Ring/SmoothingSeminorm.lean

```diff
@@ -0,0 +1,625 @@
+      obtain ⟨k, hk⟩ := hμ_lim (Set.Ioo (0 : ℝ) (1 + ε)) h1 isOpen_Ioo
+      exact hc_bd (1 + ε) k fun b hb => le_of_lt (Set.mem_Ioo.mp (hk b hb)).2
+
+theorem tendsto_smoothingFun_comp (hμ1 : μ 1 ≤ 1) (x : R) {ψ : ℕ → ℕ}
+    (hψ_mono : StrictMono ψ) :
+    Tendsto (fun n : ℕ => μ (x ^ ψ n) ^ (1 / ψ n : ℝ)) atTop (𝓝 (smoothingFun μ x)) :=
+  haveI hψ_lim' : Tendsto ψ atTop atTop := StrictMono.tendsto_atTop hψ_mono
+  (tendsto_smoothingFun_of_map_one_le_one μ hμ1 x).comp hψ_lim'
+
+private theorem limsup_mu_le (hμ1 : μ 1 ≤ 1) {s : ℕ → ℕ} (hs_le : ∀ n : ℕ, s n ≤ n) {x : R}
+    {a : ℝ} (a_in : a ∈ Set.Icc (0 : ℝ) 1) {ψ : ℕ → ℕ} (hψ_mono : StrictMono ψ)
+    (hψ_lim : Tendsto ((fun n : ℕ => (s n : ℝ) / ↑n) ∘ ψ) atTop (𝓝 a)) :
+    limsup (fun n : ℕ => μ (x ^ s (ψ n)) ^ (1 / (ψ n : ℝ))) atTop ≤ smoothingFun μ x ^ a := by
+  by_cases ha : a = 0
+  · rw [ha] at hψ_lim
+    calc limsup (fun n : ℕ => μ (x ^ s (ψ n)) ^ (1 / (ψ n : ℝ))) atTop ≤
+          limsup (fun n : ℕ => μ x ^ ((s (ψ n) : ℝ) * (1 / (ψ n : ℝ)))) atTop := by
+          apply csInf_le_csInf _ (μ_nonempty μ hs_le ψ)
+          · intro b hb
+            simp only [eventually_map, eventually_atTop, ge_iff_le, Set.mem_setOf_eq] at hb ⊢
+            obtain ⟨m, hm⟩ := hb
+            use m
+            intro k hkm
+            apply le_trans _ (hm k hkm)
+            rw [rpow_mul (apply_nonneg μ x), rpow_natCast]
+            exact rpow_le_rpow (apply_nonneg μ _) (map_pow_le_pow' hμ1 x _)
+              (one_div_nonneg.mpr (cast_nonneg _))
+          · use 0
+            simp only [mem_lowerBounds, eventually_map, eventually_atTop, ge_iff_le,
+              Set.mem_setOf_eq, forall_exists_index]
+            exact fun _ m hm ↦ le_trans (rpow_nonneg (apply_nonneg μ _) _) (hm m (le_refl _))
+      _ ≤ 1 := (μ_limsup_le_one μ hs_le hψ_lim)
+      _ = smoothingFun μ x ^ a := by rw [ha, rpow_zero]
+  · have ha_pos : 0 < a := lt_of_le_of_ne a_in.1 (Ne.symm ha)
+    have h_eq : (fun n : ℕ =>
+        (μ (x ^ s (ψ n)) ^ (1 / (s (ψ n) : ℝ))) ^ ((s (ψ n) : ℝ) / (ψ n : ℝ))) =ᶠ[atTop]
+        fun n : ℕ => μ (x ^ s (ψ n)) ^ (1 / (ψ n : ℝ)) := by
+      have h : (fun n : ℕ => (1 : ℝ) / (s (ψ n) : ℝ) * (s (ψ n) : ℝ)) =ᶠ[atTop] 1 := by
+        apply Filter.EventuallyEq.div_mul_cancel_atTop
+        exact Tendsto.num (tendsto_natCast_atTop_atTop.comp hψ_mono.tendsto_atTop) ha_pos hψ_lim
+      simp_rw [← rpow_mul (apply_nonneg μ _), mul_div]
+      exact EventuallyEq.comp₂ EventuallyEq.rfl HPow.hPow (h.div EventuallyEq.rfl)
+    exact le_of_eq (Tendsto.limsup_eq (Tendsto.congr' h_eq
+      ((((tendsto_smoothingFun_of_map_one_le_one μ hμ1 x).comp
+      ((tendsto_natCast_atTop_iff (R := ℝ)).mp <|
+        Tendsto.num (tendsto_natCast_atTop_atTop.comp hψ_mono.tendsto_atTop)
+          ha_pos hψ_lim)).rpow hψ_lim (Or.inr ha_pos)))))
+
+/-- If `μ 1 ≤ 1` and `μ` is nonarchimedean, then `smoothingFun μ` is nonarchimedean. -/
+theorem isNonarchimedean_smoothingFun (hμ1 : μ 1 ≤ 1) (hna : IsNonarchimedean μ) :
```

Comment:
The proof is quite long, can you comment it a little bit, either before or inline?

Reply:
I have added some inline comments.

Reply:
Thanks!